File: sdks/go/inbox.go
package chakramcp

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"sync"
	"time"
)

type InboxClient struct {
	c *Client
}

// Pull atomically claims the oldest pending invocations targeting an
// agent you own. Concurrent pullers (across machines) get disjoint
// batches via FOR UPDATE SKIP LOCKED at the DB.
func (in *InboxClient) Pull(ctx context.Context, agentID string, limit int) ([]Invocation, error) {
	q := url.Values{}
	q.Set("agent_id", agentID)
	if limit > 0 {
		q.Set("limit", strconv.Itoa(limit))
	}
	var out []Invocation
	if err := in.c.do(ctx, http.MethodGet, in.c.relayURL, "/v1/inbox?"+q.Encode(), nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// Respond reports the result for an in_progress invocation.
func (in *InboxClient) Respond(ctx context.Context, invocationID string, result HandlerResult) (*Invocation, error) {
	var out Invocation
	if err := in.c.do(ctx, http.MethodPost, in.c.relayURL, "/v1/invocations/"+url.PathEscape(invocationID)+"/result", result, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Handler is the user-supplied function dispatched per invocation by
// Serve. Return Succeeded(value) or Failed("reason") (or set the
// HandlerResult fields directly). Returning an error reports the
// invocation as failed with the error's message and the loop keeps
// going.
type Handler func(ctx context.Context, inv Invocation) (HandlerResult, error)

// ServeOptions configures Serve. Zero-valued fields fall back to
// PollInterval=2s, BatchSize=25.
type ServeOptions struct {
	PollInterval time.Duration
	BatchSize    int
	// OnError, if set, is called when a pull or respond fails.
	// It is also called when a handler returns an error.
	OnError func(err error, inv *Invocation)
}

// Serve runs a long-lived pull → handler → respond loop. It returns
// when ctx is cancelled. Handler errors are caught and reported as
// failed invocations; the loop keeps going.
func (in *InboxClient) Serve(ctx context.Context, agentID string, handler Handler, opts ServeOptions) error {
	if handler == nil {
		return errors.New("chakramcp: Serve requires a non-nil handler")
	}
	interval := opts.PollInterval
	if interval == 0 {
		interval = 2 * time.Second
	}
	batch := opts.BatchSize
	if batch == 0 {
		batch = 25
	}
	for {
		if err := ctx.Err(); err != nil {
			return nil
		}
		invs, err := in.Pull(ctx, agentID, batch)
		if err != nil {
			if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
				return nil
			}
			if opts.OnError != nil {
				opts.OnError(err, nil)
			}
			if !sleep(ctx, interval) {
				return nil
			}
			continue
		}
		if len(invs) == 0 {
			if !sleep(ctx, interval) {
				return nil
			}
			continue
		}
		// Process in parallel — invocations are independent.
		var wg sync.WaitGroup
		for _, inv := range invs {
			wg.Add(1)
			invCopy := inv
			go func() {
				defer wg.Done()
				in.handleOne(ctx, invCopy, handler, opts)
			}()
		}
		wg.Wait()
	}
}

func (in *InboxClient) handleOne(ctx context.Context, inv Invocation, handler Handler, opts ServeOptions) {
	defer func() {
		if r := recover(); r != nil {
			err := fmt.Errorf("handler panicked: %v", r)
			if opts.OnError != nil {
				opts.OnError(err, &inv)
			}
			_, _ = in.Respond(ctx, inv.ID, Failed(err.Error()))
		}
	}()
	res, err := handler(ctx, inv)
	if err != nil {
		if opts.OnError != nil {
			opts.OnError(err, &inv)
		}
		_, _ = in.Respond(ctx, inv.ID, Failed(err.Error()))
		return
	}
	if _, err := in.Respond(ctx, inv.ID, res); err != nil {
		if opts.OnError != nil {
			opts.OnError(err, &inv)
		}
	}
}

// sleep returns false if ctx was cancelled while sleeping.
func sleep(ctx context.Context, d time.Duration) bool {
	t := time.NewTimer(d)
	defer t.Stop()
	select {
	case <-ctx.Done():
		return false
	case <-t.C:
		return true
	}
}

File: sdks/rust/src/inbox.rs
//! Inbox client + the long-running `serve` loop.

use std::future::Future;
use std::time::Duration;

use serde_json::{json, Value};
use tokio_util::sync::CancellationToken;

use crate::client::ChakraMCP;
use crate::error::Result;
use crate::types::Invocation;

#[derive(Debug, Clone)]
pub enum HandlerResult {
    Succeeded(Value),
    Failed(String),
}

pub struct InboxClient<'a> {
    parent: &'a ChakraMCP,
}

impl<'a> InboxClient<'a> {
    pub(crate) fn new(parent: &'a ChakraMCP) -> Self {
        Self { parent }
    }

    /// Atomically claim the oldest pending invocations targeting an
    /// agent you own. Concurrent pullers (across machines) get
    /// disjoint batches via `FOR UPDATE SKIP LOCKED` at the DB.
    pub async fn pull(&self, agent_id: &str, limit: Option<u32>) -> Result<Vec<Invocation>> {
        let mut path = format!("/v1/inbox?agent_id={}", urlencode(agent_id));
        if let Some(n) = limit {
            path.push_str(&format!("&limit={n}"));
        }
        self.parent.relay_get(&path).await
    }

    pub async fn respond_succeeded(
        &self,
        invocation_id: &str,
        output: Value,
    ) -> Result<Invocation> {
        self.parent
            .relay_post(
                &format!("/v1/invocations/{}/result", urlencode(invocation_id)),
                &json!({ "status": "succeeded", "output": output }),
            )
            .await
    }

    pub async fn respond_failed(
        &self,
        invocation_id: &str,
        error: &str,
    ) -> Result<Invocation> {
        self.parent
            .relay_post(
                &format!("/v1/invocations/{}/result", urlencode(invocation_id)),
                &json!({ "status": "failed", "error": error }),
            )
            .await
    }

    /// Long-running pull → handler → respond loop. Returns a builder so
    /// you can configure and either `.await` it or attach a
    /// CancellationToken first.
    ///
    /// The handler is `FnMut(Invocation) -> Future<Output = Result<HandlerResult, _>>`.
    /// Errors raised by the handler - both Err returns and panics
    /// turned into errors - are reported as `failed` invocations and
    /// the loop continues.
    pub fn serve<F, Fut, E>(
        self,
        agent_id: impl Into<String>,
        handler: F,
    ) -> ServeBuilder<'a, F, Fut, E>
    where
        F: FnMut(Invocation) -> Fut + Send + 'a,
        Fut: Future<Output = std::result::Result<HandlerResult, E>> + Send + 'a,
        E: std::fmt::Display + Send + 'a,
    {
        ServeBuilder {
            inbox: self,
            agent_id: agent_id.into(),
            handler,
            poll_interval: Duration::from_secs(2),
            batch_size: 25,
            cancel: None,
            _phantom: std::marker::PhantomData,
        }
    }
}

pub struct ServeBuilder<'a, F, Fut, E> {
    inbox: InboxClient<'a>,
    agent_id: String,
    handler: F,
    poll_interval: Duration,
    batch_size: u32,
    cancel: Option<CancellationToken>,
    _phantom: std::marker::PhantomData<fn() -> (Fut, E)>,
}

impl<'a, F, Fut, E> ServeBuilder<'a, F, Fut, E>
where
    F: FnMut(Invocation) -> Fut + Send + 'a,
    Fut: Future<Output = std::result::Result<HandlerResult, E>> + Send + 'a,
    E: std::fmt::Display + Send + 'a,
{
    pub fn poll_interval(mut self, d: Duration) -> Self {
        self.poll_interval = d;
        self
    }
    pub fn batch_size(mut self, n: u32) -> Self {
        self.batch_size = n;
        self
    }
    pub fn with_cancellation(mut self, token: CancellationToken) -> Self {
        self.cancel = Some(token);
        self
    }

    /// Run the loop. Returns when the cancellation token fires.
    pub async fn run(mut self) -> Result<()> {
        let cancel = self.cancel.unwrap_or_default();
        loop {
            if cancel.is_cancelled() {
                return Ok(());
            }
            let batch = match tokio::select! {
                _ = cancel.cancelled() => return Ok(()),
                r = self.inbox.pull(&self.agent_id, Some(self.batch_size)) => r,
            } {
                Ok(b) => b,
                Err(_) => {
                    // Transient pull errors - sleep and retry.
                    if cancel
                        .run_until_cancelled(tokio::time::sleep(self.poll_interval))
                        .await
                        .is_none()
                    {
                        return Ok(());
                    }
                    continue;
                }
            };
            if batch.is_empty() {
                if cancel
                    .run_until_cancelled(tokio::time::sleep(self.poll_interval))
                    .await
                    .is_none()
                {
                    return Ok(());
                }
                continue;
            }
            for inv in batch {
                if cancel.is_cancelled() {
                    return Ok(());
                }
                let id = inv.id.clone();
                let result = (self.handler)(inv).await;
                match result {
                    Ok(HandlerResult::Succeeded(out)) => {
                        let _ = self.inbox.respond_succeeded(&id, out).await;
                    }
                    Ok(HandlerResult::Failed(msg)) => {
                        let _ = self.inbox.respond_failed(&id, &msg).await;
                    }
                    Err(e) => {
                        let _ = self.inbox.respond_failed(&id, &e.to_string()).await;
                    }
                }
            }
        }
    }
}

impl<'a, F, Fut, E> std::future::IntoFuture for ServeBuilder<'a, F, Fut, E>
where
    F: FnMut(Invocation) -> Fut + Send + 'a,
    Fut: Future<Output = std::result::Result<HandlerResult, E>> + Send + 'a,
    E: std::fmt::Display + Send + 'a,
{
    type Output = Result<()>;
    type IntoFuture = std::pin::Pin<Box<dyn Future<Output = Self::Output> + Send + 'a>>;

    fn into_future(self) -> Self::IntoFuture {
        Box::pin(self.run())
    }
}

fn urlencode(s: &str) -> String {
    url::form_urlencoded::byte_serialize(s.as_bytes()).collect()
}
